src/com/ms/silverking/cloud/dht/management/aws/Util.java

```diff
@@ -21,377 +21,373 @@
+      }
+
+      if (ips.isEmpty())
+        break;
+
+      checkForTimeoutException(sleepSeconds, totalRunTimeSeconds, retriesCount, "running");
+
+      sleep(sleepSeconds);
+      retriesCount++;
+      lastMinutePrinted = printMinutesElapsed(sleepSeconds, retriesCount, lastMinutePrinted);
     }
-    
-    static List<Instance> findInstancesStateWithKeyPair(AmazonEC2 ec2, String keyPair, InstanceState state) {
-        print("Finding " + state + " Instances");
-
-        List<Instance> instances = new ArrayList<>();
-        
-        DescribeInstancesRequest request = new DescribeInstancesRequest();
-        while (true) {
-            DescribeInstancesResult response = ec2.describeInstances(request);
-
-            for (Reservation reservation : response.getReservations())
-                for (Instance instance : reservation.getInstances()) {
-                    printInstance(instance);
-                    switch (state) {
-                        case STOPPED:
-                            if ( isStopped(instance) && isKeyPair(instance, keyPair) )
-                                instances.add(instance);
-                            break;
-                        case RUNNING:
-                            if ( isRunning(instance) && isKeyPair(instance, keyPair) )
-                                instances.add(instance);
-                            break;
-                        default: 
-                            throwRuntimeException("Unknown Instance state: " + state);
-                    }
-                    
-                }
-
-            if (response.getNextToken() == null)
-                break;
-            
-            debugPrint("token: " + response.getNextToken());
-            request.setNextToken(response.getNextToken());
+
+    printDone(getIps(instances));
+  }
+
+  private static void checkForTimeoutException(long sleepSeconds, long totalRunTimeSeconds, int retriesCount,
+      String status) {
+    if (retriesCount * sleepSeconds > totalRunTimeSeconds)
+      throwTimeoutException(status);
+  }
+
+  private static void throwTimeoutException(String status) {
+    throwRuntimeException("instances should have been " + status + " by now...");
+  }
+
+  static void waitForInstancesToBeReachable(AmazonEC2 ec2, List<Instance> instances) {
+    printNoDot("  Waiting for Instances to be reachable");
+    printNoDot("    reachable = BOTH (System and Instance) status checks PASS");
+    printNoDot("    be patient, it usually takes ~5mins");
+
+    DescribeInstanceStatusRequest disRequest = new DescribeInstanceStatusRequest();
+    disRequest.withInstanceIds(getInstanceIds(instances));
+
+    long lastMinutePrinted = 0;
+    long sleepSeconds = 15;
+    long totalRunTimeSeconds = 10 * 60;
+    int retriesCount = 0;
+    List<String> ips = getIps(instances);
+    while (!ips.isEmpty()) {
+      DescribeInstanceStatusResult response = ec2.describeInstanceStatus(disRequest);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code appears to be waiting for a resource before it runs. You could use the waiters feature to help improve efficiency. Consider using InstanceStatusOk or SystemStatusOk. For more information, see [https://aws.amazon.com/blogs/developer/waiters-in-the-aws-sdk-for-java/](https://aws.amazon.com/blogs/developer/waiters-in-the-aws-sdk-for-java/)